Libmesh update 20250205

roystgnr opened the pull request:
Update libMesh submodule:

- Bug fix: `normals` vectors at the sides of edge elements in 3D, which were previously hard-coded to return `(-1,0,0)` and `(1,0,0)`  at points 0 and 1 respectively (the correct solution only for x-axis-aligned edges), are now correctly calculated for all edges in 2D/3D.  Compatibility with this fix required a corresponding fix in one MOOSE application; developers of any applications not in MOOSE CI who use non-x-axis-aligned edge elements may wish to test their codes for result changes with this update.
- Bug fix: H(curl) and H(div) element calculations are now correct on 2D elements embedded in 3D space and not parallel to the xy plane.
- Bug fix: Some TIMPI communications of "packed-range" data (data with variable-size data types), when given enough input data (typically ~4MB by default) to split between multiple buffers and given an output iterator that requires correct incrementation, were not correctly preserving that iterator from one buffer to the next.  This has been fixed.
- Command line option `--mpi-thread-type` to manually choose between "single", "funneled", "serialized", and "multiple" MPI initialization threading options.  Initialization still defaults to "funneled" (for maximum compatibility with older MPI implementations), but users of slate/strumpack may need to specify "multiple" for compatibility with those packages.
- PetscMatrix support for hash-table-based matrices with newer PETSc versions.  This will allow MOOSE much more flexibility in problems (e.g. contact) where a priori determination of matrix sparsity patterns is difficult.
- Fixes for `HYPRE_SetMemoryLocation` in more complicated (e.g. field split) configurations
- Removed `contrib/boost`, which was a version incompatible with C++17.  libMesh codes (including MOOSE) which require boost now must have an external boost installation to use.
- Support for and compiler flag selection for configuring libMesh with newer Intel compilers
- `MeshTools::n_connected_components()` function to count the number of connected components in a mesh
- Added operator[] for DenseVector.  This is more consistent with other vector/map container types, and enables more compatibility with generic templated algorithms.
- `PetscMatrix::frobenius_norm()` method, and refactoring of other norm calculations
- Improvements and more options in `calculator` app
- Improvements and bug fixes in reduced-basis code
- New grad(div()) problem example code with Raviart-Thomas elements
- Support for "packing" for Cap'n'Proto users
- Bug fix in FParser optimized JIT output
- Bug fix for corrupt time stamps in Exodus outputs with a single time step
- Bug fixes in programmatic setting of PETSc preconditioners
- Workarounds for older PETSc error-checking macros are now removed; raw PETSc code should use `LibmeshPetscCall` for error checking.
- General cleanup in autoconf (`./configure` source) code
- General refactoring in `FEInterface`

Edit: a force push also adds:
- Fix for a multithreading error in applications where multiple threads are encountering elements with negative mapping Jacobians
- Preconditioning option, to scale new raw degrees of freedom, when importing mesh constraint rows from a sparse matrix.  This can greatly improve solver robustness on such constrained systems.
- New `transient_ex3` example application demonstrating a explicit DG/FV formulation of the 2D advection equation

roystgnr committed:
Newsletter entries for libMesh submodule update

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 6ef7d4395794104f48dae1fd48e64077207188e8
+Subproject commit 60bc66e8ec77774a925123e5edc3e06776280517
```
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2675659/), step Conda build config check on 74b318c wanted to post the following:

A change of the following file(s) triggered this check:

  conda/libmesh

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2675659/), step Conda libmesh build config check on 74b318c wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The libmesh submodule or configuration was changed but the conda build config was not

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2675659/), step Conda libmesh build config notification on 74b318c wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2675655/), step Docs: sync website on 74b318c wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/29817/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2675702/), step Generate coverage on 74b318c wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>e5db19</th>
      <th colspan=4>#29817 74b318</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.22%</td>
      <td>85.21%</td>
      <td>-0.01%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>7997</td>
      <td>7996</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1387</td>
      <td>1388</td>
      <td>+1</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29817/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29817/coverage/fluid_properties)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/29817/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad commented:
It looks like I may need to go through some apps and fix up the PETSc macros

lindsayad commented:
- [x] Griffin patch: https://github.inl.gov/ncrc/griffin/pull/2327
- [x] Subchannel patch: https://github.inl.gov/ncrc/subchannel/pull/351

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 6ef7d4395794104f48dae1fd48e64077207188e8
+Subproject commit 5b27fa72ab898444fd239cb345e5dfd910905108
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2692236/), step Conda build config check on 9622db6 wanted to post the following:

A change of the following file(s) triggered this check:

  conda/libmesh

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2692236/), step Conda libmesh build config check on 9622db6 wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The libmesh submodule or configuration was changed but the conda build config was not

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2692236/), step Conda libmesh build config notification on 9622db6 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

lindsayad commented:
I think that when https://github.inl.gov/ncrc/pronghorn/pull/1292 and https://github.com/idaholab/virtual_test_bed/pull/520 are merged, then pronghorn devel (which currently has a subchannel module update that fixes the libmesh petsc call macros) can go to master, and then I believe we'll be all set to go for invalidating the failing app tests

moosebuild commented:
Job [Documentation internal apps](https://civet.inl.gov/job/2692234/) on 9622db6 : invalidated by @lindsayad

moosebuild commented:
Job [Internal app tests](https://civet.inl.gov/job/2692267/) on 9622db6 : invalidated by @lindsayad

lindsayad commented:
relap-7 and sockeye should be fixed now thanks to @joshuahansel. Looks like we need a rebase here

roystgnr commented:
I'll do the rebase in a sec.

roystgnr commented:
Well, I'll rebase the newsletter, anyway; simpler to just redo the other commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2699262/), step Conda build config check on 26b9557 wanted to post the following:

A change of the following file(s) triggered this check:

  conda/libmesh

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The conda build config was not changed but one or more of its dependencies have changed

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2699262/), step Conda libmesh build config check on 26b9557 wanted to post the following:

A change of the following file(s) triggered this check:

  libmesh

The following file(s) are unchanged:

  conda/mpi/conda_build_config.yaml

The libmesh submodule or configuration was changed but the conda build config was not

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2699262/), step Conda libmesh build config notification on 26b9557 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

lindsayad commented:
I added the other targets. Looks like we need another rebase

milljm commented:
neat... you'll be one of the first to follow: https://github.com/idaholab/moose/wiki/Updating-packages

roystgnr committed:
Update the docs for the update

roystgnr committed:
Update libMesh submodule

- Bug fix: `normals` vectors at the sides of edge elements in 3D,
  which were previously hard-coded to return `(-1,0,0)` and `(1,0,0)`
  at points 0 and 1 respectively (the correct solution only for
  x-axis-aligned edges), are now correctly calculated for all edges in
  2D/3D.  Compatibility with this fix required a corresponding fix in
  one MOOSE application; developers of any applications not in MOOSE
  CI who use non-x-axis-aligned edge elements may wish to test their
  codes for result changes with this update.
- Bug fix: H(curl) and H(div) element calculations are now correct on
  2D elements embedded in 3D space and not parallel to the xy plane.
- Bug fix: Some TIMPI communications of "packed-range" data (data with
  variable-size data types), when given enough input data (typically
  ~4MB by default) to split between multiple buffers and given an
  output iterator that requires correct incrementation, were not
  correctly preserving that iterator from one buffer to the next.
  This has been fixed.
- Fix for a multithreading error in applications where multiple
  threads are encountering elements with negative mapping Jacobians
- Preconditioning option, to scale new raw degrees of freedom, when
  importing mesh constraint rows from a sparse matrix.  This can
  greatly improve solver robustness on such constrained systems.
- Support for the combination of subdomain-restricted variables with
  subdomain-spanning mesh constraints.  This fixes errors from certain
  use cases of subdomain-restricted variables on IsoGeometric
  Analysis meshes.
- Command line option `--mpi-thread-type` to manually choose between
  "single", "funneled", "serialized", and "multiple" MPI
  initialization threading options.  Initialization still defaults to
  "funneled" (for maximum compatibility with older MPI
  implementations), but users of slate/strumpack may need to specify
  "multiple" for compatibility with those packages.
- PetscMatrix support for hash-table-based matrices with newer PETSc
  versions.  This will allow MOOSE much more flexibility in problems
  (e.g. contact) where a priori determination of matrix sparsity
  patterns is difficult.
- Fixes for `HYPRE_SetMemoryLocation` in more complicated (e.g. field
  split) configurations
- Removed `contrib/boost`, which was a version incompatible with
  C++17.  libMesh codes (including MOOSE) which require boost now must
  have an external boost installation to use.
- Support for and compiler flag selection for configuring libMesh with
  newer Intel compilers
- Bug fixes and first-order two-dimensional test coverage for
  VariationalMeshSmoother.
- `MeshTools::n_connected_components()` function to count the number
  of connected components in a mesh
- Added operator[] for DenseVector.  This is more consistent with
  other vector/map container types, and enables more compatibility
  with generic templated algorithms.
- `PetscMatrix::frobenius_norm()` method, and refactoring of other
  norm calculations
- New `transient_ex3` example application demonstrating a explicit
  DG/FV formulation of the 2D advection equation
- Improvements and more options in `calculator` app
- Improvements and bug fixes in reduced-basis code
- New grad(div()) problem example code with Raviart-Thomas elements
- Support for "packing" for Cap'n'Proto users
- Bug fix in FParser optimized JIT output
- Bug fix for `elem_default_orders()` print style in
  `MeshBase::get_info()` and `print_info()`.
- Bug fix for corrupt time stamps in Exodus outputs with a single time
  step
- Bug fixes in programmatic setting of PETSc preconditioners
- Workarounds for older PETSc error-checking macros are now removed;
  raw PETSc code should use `LibmeshPetscCall` for error checking.
- General cleanup in autoconf (`./configure` source) code
- General refactoring in `FEInterface`

Refs #0

moosebuild commented on libmesh:
```diff
@@ -1 +1 @@
-Subproject commit 6ef7d4395794104f48dae1fd48e64077207188e8
+Subproject commit b30db56c835158b7f36397e4339c493a28592480
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2712878/), step Versioner verify on e8eb234 wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status    | hash    | to hash   | version            | to version         |
|-----------|-----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE OK | b8b3d69 | dabc922   | 2024.12.02=mpich_1 | 2025.02.25=mpich_0 |
| moose-dev | CHANGE OK | 5b49120 | 026ca42   | 2025.02.18=mpich   | 2025.02.25=mpich   |

lindsayad commented:
Anyway we can wait to get libMesh/libmesh#4091 in here?

roystgnr commented:
Okay, most of those latest set of failures were "our CI can't always complete a TCP connection" issues, and I've restarted those, but we've also got a framework test failure in https://civet.inl.gov/job/2712966/ that looks real (related to https://github.com/idaholab/moose/pull/28785 ?  But I don't see how it would have passed there) and a Bison test failure in https://civet.inl.gov/job/2712965/ that just looks bizarre - it's trying to open a `CheckpointIO` file with an `ExodusII_IO` reader??

moosebuild commented:
Job [Apptainer openmpi internal apps](https://civet.inl.gov/job/2712965/) on e8eb234 : invalidated by @lindsayad

lindsayad commented:
I'm pretty sure the latter failure is because of openmpi file system issues that we've seen in many places. I'm invalidating to see what happens

lindsayad reviewed (approved):
@milljm do the conda changes look good to you?

milljm reviewed (approved):
The output of the verification looks great to me. My only concern is another update that might vie for merge rights affecting moose-dev's version: https://github.com/idaholab/moose/pull/29930

lindsayad commented:
Oh yea sure enough

lindsayad commented:
Maybe this means we can get libmesh/libmesh#4091 in 😆

roystgnr committed:
Update versioner templates

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2720205/), step Versioner verify on 40e6bda wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 0 failed

| package   | status    | hash    | to hash   | version            | to version         |
|-----------|-----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE OK | b8b3d69 | dabc922   | 2024.12.02=mpich_1 | 2025.02.25=mpich_0 |
| moose-dev | CHANGE OK | 5b49120 | 026ca42   | 2025.02.18=mpich   | 2025.02.25=mpich   |

roystgnr commented:
I'll start another submodule update PR next week, even if all it's got is 4091 and the Prism+subdomain-restriction+AMR fix and the FE+Quadrature API updates ... but at this point I don't want to add any additional uncertainty to this one.  Taking 3 weeks to get a monthly submodule update in is embarrassing enough already.

BWXT is still working with hand-compiled software mixes right now, to work around an IGA conditioning issue with a fix in libMesh.  I'd love to be able to answer Coreform's reasonable question of "do you have a timeline for when we'll be able to run the IGA workflow on MOOSE master?" with something other than "Nope!", but at least I can avoid increasing the delay.
 
In e.g. https://civet.inl.gov/job/2719589/ we're seeing the same ARM Mac test failures (hopefully to be fixed permanently when #29962 hits devel) from other PRs, so at least there doesn't seem to be too much uncertainty on that front.  I just kicked the tests here (before the re-versioning force pushes) and I saw the quicker recipe passing already, so it really must be come kind of parallel race condition.

And ... the re-versioning force pushes didn't actually fix the branch conflicts.  If I rebase on origin/devel, we've still got a conflict, and if I rebase on origin/next then `./scripts/versioner.py --build-templates` thinks I haven't changed `moose-dev` ... which can't be correct, can it?  I'll try pushing from that anyway and let @milljm tell me.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2720304/), step Versioner verify on c1c8fab wanted to post the following:

### Versioner templates

Found 16 templates, 0 failed


### Versioner versions

Found 10 packages, 2 changed, 1 failed

| package   | status    | hash    | to hash   | version            | to version         |
|-----------|-----------|---------|-----------|--------------------|--------------------|
| libmesh   | CHANGE OK | b8b3d69 | dabc922   | 2024.12.02=mpich_1 | 2025.02.25=mpich_0 |
| moose-dev | NEED BUMP | 98494d1 | a90351b   | 2025.02.25=mpich   | 2025.02.25=mpich   |

Verification failed.

roystgnr commented:
Yeah, neither rebase works.

roystgnr commented:
I was hoping next->devel would finish up last night, but no; the API change mixed with our fix in #29962 broke SAM builds in e.g. https://civet.inl.gov/job/2718600/

loganharbour commented:
@roystgnr I'll look at this shortly. Please keep it as is. I want to make sure our failure checking works.

We've got two separate moose-dev bumps stuck on next, and this has an old date so it's perfect for sanity checking.

loganharbour committed:
Update versioner hashes

milljm reviewed (approved):
bump looks good to me! 👍

lindsayad commented:
I don't understand why the openmpi tests are taking so long. Can we just merge this bad boy?

roystgnr commented:
2 crashes with stack traces in PetscInitialize, one timeout just starting a test.

My vote is for "just merge this bad boy" - I could make a long list of all the things I could have theoretically broken with the dozen odd major changes in this update, but "PetscInitialize, but only with OpenMPI" would not be on it.

It's disappointing to see OpenMPI so flaky, though.  Way back when I used to switch between OpenMPI and MPICH frequently, well, neither of them was perfect, but OpenMPI was always the more reliable of the two.

lindsayad commented:
I would still like to remove the openmpi target, at least on PRs. The sum of time developers have spent wondering about false positives has to at least be on the order of hours at this point

roystgnr commented:
I like the OpenMPI target as a "sanity check" for weird MPI-looking failures in MPICH, but it really should be going yellow on failure, not red.

What's the verdict for this case, though?  Merge despite OpenMPI, or invalidate and cross our fingers?

roystgnr commented:
Ah, wait, never mind, we're still branch locked.
